singularity-eos/closure/mixed_cell_models.hpp

```diff
@@ -155,6 +155,12 @@ class CacheAccessor {
 
 template <typename EOSIndexer, typename RealIndexer>
 class PTESolverBase {
+  friend class PTESolverRhoT;
+  friend class PTESolverRhoU;
+  friend class PTESolverRhoU;
+  friend class PTESolverFixedP;
+  friend class PTESolverFixedT;
```

Comment:
I really dislike using `friend class` here as it mixes inheritance access in a way that I worry is hard to disentangle. For example, having the base class know about its children with `friend` means that a downstream code can't build on our solver infrastructure to add their own solver with our base functionality. They have to make a whole new class, since they can't add "friendship" downstream. It's also just one more place things have to be modified when a new class is added.

(Big picture, I think `friend` is a language feature that should be used very sparingly, as it breaks encapsulation. It essentially means all the classes/functions that are `friends` are part of a single monolithic implementation. Sometimes this is the right thing to do. But when faced with that design choice, I think it's worth thinking carefully about alternatives.)

That said, it's actually not clear to me: why is this necessary here? Maybe I'm missing something about the inheritance structure.

Reply:
I want to emphasize that these changes were made to duplicate the functionality that already exists through the `using` statements we already had.

> I really dislike using friend class here as it mixes inheritance access in a way that I worry is hard to disentangle.

Is this worse than mixing data access through `using` statements?

> ...having the base class know about its children with friend means that a downstream code can't build on our solver infrastructure to add their own solver with our base functionality

They can still adopt the `using` pattern that we currently have and accomplish the same thing. This change only _enables_ data access through inheritance.

> Big picture, I think friend is a language feature that should be used very sparingly, as it breaks encapsulation. It essentially means all the classes/functions that are friends are part of a single monolithic implementation.

Maybe I'm failing to grasp some of the nuances here, but I'd argue that our current implementation already does this.


I'm certainly open to alternatives, but to me (and again, I'm new to C++ so I may be completely wrong) the `using` statements feel like a hack to expose variables to a different scope.

Reply:
I tend to agree with @Yurlungur. While neither solution is optimal, I'd prefer the `using` statements over `this->`. As my simple mind (and procedural programming tendencies) sees things, the usage of a base class is to reduce code duplication. Both of these implementations accomplish this task, but I find the `using` route to be more readable and the absence of `friend` classes an additional benefit as it communicates to downstream projects that one must register their implementations with singularity-eos in order to "correctly" inherit our base class's functionality, which is not something we want to communicate IMO.

Reply:
Discussing with @jhp-lanl offline---I'm actually confused why the `friend` stuff is needed. I can't reproduce in some simple examples.

We also discovered that the `this` pointer is only required because of the templates (which is my memory of the language). This has to do with implementation details of how unqualified name lookup interacts with templates in the language. From the C++11 standard:

> In the definition of a class or class template, if a base class depends on a template-parameter, the base class scope is not examined during unqualified name lookup either at the point of definition of the class template or member or during an instantiation of the class template or member. 

To my mind, this makes both `using` and `this` essentially equivalent "hacks" around the failure of the language to properly do unqualified name lookup. So I think it's just a question of which we as a team prefer stylistically. (I feel pretty strongly I don't want `friend` classes here... but maybe we can work around that if we decide to go with the `this` pointer.)

Reply:
I removed the `friend` stuff because it isn't needed. I _thought_ it was needed at some point, but that was my misunderstanding.

I agree on the `using` vs `this->`. Given that we have two votes against, I'm inclined to scrap it.